### DISCUSSION TITLE
Help: How can I run Storybook 7+ inside an Angular library

### DISCUSSION THREAD

**Discussion by AE1NS** [storybookjs/storybook#20808] [Help] [answered]
The [documentation](https://storybook.js.org/docs/7.0/angular/get-started/install) for Angular/Storybook 7+ tells me to run storybook with `ng run my-project:storybook`. But how can I install storybook to an angular library project, where no `"projectType": "application"` exists inside the angular.json? Isnt there another way to set the builder for storybook?

When I run `storybook dev` I get the following output:
```
info => Starting manager..
info Addon-docs: using MDX2
info => Using implicit CSS loaders
ERR! Your Storybook startup uses a solution that is not supported.
ERR! You must use angular builder to have an explicit configuration on the project used in angular.json
ERR! Read more at:
ERR! - https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#sb-angular-builder)
ERR! - https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#angular13)
WARN Force closed manager build
ERR! Error: angularBrowserTarget is undefined.
ERR!     at checkForLegacyBuildOptions (c:\_entwicklung\aeinsshared\node_modules\.pnpm\@storybook+angular@7.0.0-beta.35_de733f6cpddyfptp7o2fcstyxy\node_modules\@storybook\angular\dist\server\framework-preset-angular-cli.js:84:11)
ERR!     at Object.webpackFinal (c:\_entwicklung\aeinsshared\node_modules\.pnpm\@storybook+angular@7.0.0-beta.35_de733f6cpddyfptp7o2fcstyxy\node_modules\@storybook\angular\dist\server\framework-preset-angular-cli.js:16:5)
ERR!     at C:\_entwicklung\aeinsshared\node_modules\.pnpm\@storybook+core-common@7.0.0-alpha.57_yalvw3r2waubxycyb7k7qsruca\node_modules\@storybook\core-common\dist\index.js:12:617
ERR!  Error: angularBrowserTarget is undefined.
ERR!     at checkForLegacyBuildOptions (c:\_entwicklung\aeinsshared\node_modules\.pnpm\@storybook+angular@7.0.0-beta.35_de733f6cpddyfptp7o2fcstyxy\node_modules\@storybook\angular\dist\server\framework-preset-angular-cli.js:84:11)
ERR!     at Object.webpackFinal (c:\_entwicklung\aeinsshared\node_modules\.pnpm\@storybook+angular@7.0.0-beta.35_de733f6cpddyfptp7o2fcstyxy\node_modules\@storybook\angular\dist\server\framework-preset-angular-cli.js:16:5)
ERR!     at C:\_entwicklung\aeinsshared\node_modules\.pnpm\@storybook+core-common@7.0.0-alpha.57_yalvw3r2waubxycyb7k7qsruca\node_modules\@storybook\core-common\dist\index.js:12:617
```

**Comment by bzh2610** [comment, marked as answer]
You could add a run configuration inside of your angular.json which will only be used for storybook ?

```
"my-custom-library": {
[...]
      "projectType": "library",
      "architect": {
[...]
        "serve-storybook": {
          "builder": "@angular-devkit/build-angular:browser",
          "options": {
            "tsConfig": ".storybook/tsconfig.json",
            "assets": [  ],
            "index": "src/index.html",
            "main": "src/main.ts",
            "outputPath": "../app/src/main/webapp/",
            "styles": [],
          }
        },
        "storybook": {
          "builder": "@storybook/angular:start-storybook",
          "options": {
            "configDir": ".storybook",
            "browserTarget": "my-custom-library:serve-storybook",
            "compodoc": false,
            "port": 6006,
            "styles": []
          }
        }
}
```

**Reply by AE1NS** [reply]
Well, thats really easy and working. Thank you!

**Reply by 87Dev** [reply]
Hello @AE1NS 

Could you go through the entire configuration file to check it? I'm in the same situation and I can't get it to start.

Greetings

**Reply by paulo101977** [reply]
> You could add a run configuration inside of your angular.json which will only be used for storybook ?
> 
> ```
> "my-custom-library": {
> [...]
>       "projectType": "library",
>       "architect": {
> [...]
>         "serve-storybook": {
>           "builder": "@angular-devkit/build-angular:browser",
>           "options": {
>             "tsConfig": ".storybook/tsconfig.json",
>             "assets": [  ],
>             "index": "src/index.html",
>             "main": "src/main.ts",
>             "outputPath": "../app/src/main/webapp/",
>             "styles": [],
>           }
>         },
>         "storybook": {
>           "builder": "@storybook/angular:start-storybook",
>           "options": {
>             "configDir": ".storybook",
>             "browserTarget": "my-custom-library:serve-storybook",
>             "compodoc": false,
>             "port": 6006,
>             "styles": []
>           }
>         }
> }
> ```

I tried a similar configuration here to build the storybook, but it doesn't work.  The builder simply does nothing.
Edited: I found the solution! And it's simple: if "compodoc": true, or no "compodoc" in build configuration, the build doesn't happen!!!!

```
"build-storybook": {
        "builder": "@storybook/angular:build-storybook",
        "options": {
          "configDir": "projects/my-lib/.storybook",
          "browserTarget": "my-lib:build",
          "compodoc": true,
          "compodocArgs": [
            "-e",
            "json",
            "-d",
            "projects/my-lib"
          ],
          "outputDir": "storybook-static"
        }
      }
```

**Comment by Nikhil936** [comment]
I added Storybook inside a library. Angular storybook is running successfully in my project, But view is not rendering. when I check the example components which storybook has provided, I am getting below error. 

Cannot read properties of undefined (reading 'selector')
TypeError: Cannot read properties of undefined (reading 'selector')
    at computesTemplateFromComponent (http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:81513:28)
    at prepareMain (http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82562:82)
    at http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82548:17
    at http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82542:14
    at cleanArgsDecorator (http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82591:10)
    at http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82541:28
    at http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82542:14
    at withOutline (http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:81256:43)
    at hookified (http://localhost:6008/sb-preview/runtime.js:84:10607)
    at http://localhost:6008/vendors-node_modules_storybook_addon-docs_angular_index_js-node_modules_storybook_addon-essen-79ded4.iframe.bundle.js:82541:28

**Reply by valentinpalkovic** [reply]
@Nikhil936 Can you provide a reproduction? https://storybook.new
